Parameterize metric/label name validation scheme

juliusmh opened the pull request:
#### What this PR does

Direct and indirect references to the global name validation scheme were removed in favor of a per-tenant override.

- Distributors have a validation middleware that ensures metric and label names are valid.
- Rulers validate rule(group)s using this naming scheme.
- Queriers use UTF8 validation *everywhere*.

TODO:

- [x] Reach consensus on whether `streamingpromqlcompat.NameValidatingEngine` is the right approach: Decided to do UTF8 validation in query path.
- [x] Get Alertmanager updated with fix, stop using our own fork


#### Which issue(s) this PR fixes or relates to

Depends on
- https://github.com/grafana/mimir-prometheus/pull/946
- https://github.com/grafana/prometheus-alertmanager/pull/116

Fixes: #11503

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`. If changelog entry is not needed, please add the `changelog-not-needed` label to the PR.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

juliusmh commented on pkg/mimirtool/rules/rules.go:
```diff
@@ -255,7 +256,8 @@ func (r RuleNamespace) Validate(groupNodes []rulefmt.RuleGroupNode) []error {
 	for i, r := range g.Rules {
-		for _, err := range r.Validate(node.Rules[i]) {
+		// TODO(juliusmh):
+		for _, err := range r.Validate(node.Rules[i], validation.LegacyNamingScheme) {
```
Q: How should we handle validation checks in mimirtool?

aknuds1 reviewed (commented):
OK, gave it a quick review. Nice work! Please check my questions :)

juliusmh commented on pkg/mimir/modules.go:
```diff
@@ -827,6 +827,8 @@ func (t *Mimir) initQueryFrontendTripperware() (serv services.Service, err error
 		panic(fmt.Sprintf("invalid config not caught by validation: unknown PromQL engine '%s'", t.Cfg.Querier.QueryEngine))
 	}
 
+	eng = streamingpromqlcompat.NameValidatingEngine(eng, t.Overrides)
```
I decided to go with a wrapper instead, mainly because it means less changes to the query path and it's easier to implement. This also drove the decision to add name validation scheme to `promql.QueryOpts` rather than `promql.Engine`.

At the moment queriers/frontends use a single engine to process all incoming queries. In Mimir, we use one engine in query frontends, another one is used by queriers. The alternative is to pass a `QueryEngineFactory` instead of a `QueryEngine` and either:

- create a new engine for each request _or_
- create legacy- and utf8-engines and choose the appropriate one

charleskorn replied:
> @juliusmh has made me aware that, on the other hand, there are certain parts of the query engine (label replace?) that need to use the same validation scheme as the ingest path.

This is correct. `label_replace` and `label_join` will need to use the same validation scheme. 

> I also think that for the most part, queries shouldn't enforce the legacy naming scheme. My reasoning is that if the user at one point has ingested metrics with the less restrictive UTF-8 naming scheme, and then flips the switch back to the "legacy" naming scheme, why should the query engine fail queries against the metrics ingested with the UTF-8 naming scheme

I'm not aware of anywhere apart from `label_replace` and `label_join` that will enforce label name validation, so I don't believe this is an issue - is there something I've missed?

aknuds1 replied:
@charleskorn I just want to make sure we don't introduce configurable metric/label name validation in the query engine except for where we need to (`label_replace` and `label_join` as I understand it). At most we should verify that names are valid UTF-8 (apart from the noted exceptions).

juliusmh replied:
What about `count_values`, it returns a series with the aggregation label?

charleskorn replied:
Yep, good catch, you're correct.

juliusmh commented on pkg/streamingpromql/compat/name_validating_engine.go:
```diff
@@ -0,0 +1,88 @@
+		return "", err
+	}
+	for _, tenantID := range tenantIDs {
+		if e.limits.NameValidationScheme(tenantID) == prom_validation.LegacyNamingScheme {
+			return prom_validation.LegacyNamingScheme, nil
+		}
```
This PR tries to maintain the current behavior, and currently, we validate label names in `label_join`, `label_replace` and `count_values` with legacy scheme. Queries that don't use those functions support UTF8, today. This can be seen in the [tests](https://github.com/grafana/mimir/pull/11855) I did. 

The question we have to answer: is there any scenario in which data returned by a querier or frontend is fed back to mimir without validation (e.g. to ingesters directly, or with label validation disabled)? If the answer is no, we can safely remove the validation.

charleskorn replied:
I believe that data will always be validated when written back by rulers. They don't skip validation.

However, I believe we should validate names in at least `label_replace` and `label_join`, as not validating may lead to unexpected results:

* functions that consume the results of `label_replace` and `label_join` may have undefined behaviour with invalid label names
* a user testing a recording or alerting rule as a query may not realise that their rule will later fail to write samples when evaluated by rulers

juliusmh commented on pkg/cardinality/request_test.go:
```diff
@@ -98,7 +99,7 @@ func TestDecodeLabelValuesRequest(t *testing.T) {
 		require.NoError(t, err)
 
-		actual, err := DecodeLabelValuesRequest(req, 0)
+		actual, err := DecodeLabelValuesRequest(req, 0, validation.LegacyNamingScheme)
```
Mainly because at the moment, Mimir does validate this. However, you're correct that we will not be able to access those metrics after switching from utf8 to legacy. I altered package cardinality and querier to use utf8 validation.

aknuds1 commented on pkg/cardinality/request.go:
```diff
@@ -262,10 +263,10 @@ func extractLabelNames(values url.Values) ([]model.LabelName, error) {
 	for _, labelNameParam := range labelNamesParams {
-		labelName := model.LabelName(labelNameParam)
-		if !labelName.IsValid() {
+		if !validation.UTF8NamingScheme.IsValidLabelName(labelNameParam) {
```
I am wondering in this case whether this code should use a parameter (instead of hard coding), but not let it be user configurable since, as I explain above, I'm not sure we want to enforce the legacy naming scheme in queries. WDYT @charleskorn?

charleskorn replied:
I don't know enough about this code path to know what makes sense here. Does Prometheus apply validation to the label names here as well?

juliusmh replied:
The `api/v1/cardinality/*` endpoints are Mimir specific. 

The closest is probably db stats in prometheus admin API (calculated [here](https://github.com/grafana/mimir-prometheus/blob/6693a913b24cca33b98a2d98f1f954d8415a4583/tsdb/index/postings.go#L198)), where you can get some analysis on label name/value cardinality, which, as far as I can see, doesn't involve any validation. But I'm not super confident in that analysis.

juliusmh commented on pkg/streamingpromql/operators/aggregations/count_values_test.go:
```diff
@@ -222,7 +223,16 @@ func TestCountValues_GroupLabelling(t *testing.T) {
+				testCase.without,
+				memoryConsumptionTracker,
+				posrange.PositionRange{},
+				validation.LegacyNamingScheme,
```
Are we talking about this unit test or the `count_values` function in general? At runtime, MQE will use the injected/per-tenant scheme (this is just hardcoded in the test)

aknuds1 commented on pkg/querier/cardinality_analysis_handler_test.go:
```diff
@@ -705,10 +705,6 @@ func TestLabelValuesCardinalityHandler_ParseError(t *testing.T) {
 				url:                  "/label_values",
 				expectedErrorMessage: "'label_names[]' param is required",
 			},
-			"label_names param is invalid": {
-				url:                  "/label_values?label_names[]=olá",
-				expectedErrorMessage: "invalid 'label_names' param 'olá'",
-			},
```
The test case just needs an adjustment, to a string which isn't valid UTF-8:
```suggestion
			"label_names param is invalid": {
				url:                  "/label_values?label_names[]=\xff\xfe",
				expectedErrorMessage: "invalid 'label_names' param '\xff\xfe'",
			},
```

aknuds1 reviewed (changes requested):
I think this is starting to look good. We should keep that test case though.

juliusmh commented on pkg/streamingpromql/planning.go:
```diff
@@ -441,7 +441,7 @@ func findFunction(name string) (functions.Function, bool) {
 func (e *Engine) Materialize(ctx context.Context, plan *planning.QueryPlan, queryable storage.Queryable, opts promql.QueryOpts) (promql.Query, error) {
 	if opts == nil {
-		opts = promql.NewPrometheusQueryOpts(false, 0)
+		opts = promql.NewPrometheusQueryOpts(false, 0, model.UTF8Validation)
```
> @juliusmh Should q.nameValidationScheme be used here?

What is `q` ? At this point we don't have a query yet. We could give the engine `e` a configurable default. FYI, prometheus uses global `model.NameValidationScheme` here.

juliusmh commented:
As discussed @aknuds1 :
- based on r352
- Synced with https://github.com/grafana/mimir/pull/12109
- Fixed failing tests

@charleskorn :
> Could you elaborate? It seems like adding it to promql.QueryOpts would be a more natural fit rather than creating a wrapper like this (at least for Mimir, not sure about Prometheus).

For promql, the main problem is that query opts are directly constructed and passed to query engine in prometheus' [v1.NewApi](https://github.com/grafana/mimir/blob/be35f7d5d9b8a0345ad7a16eb938928523b92e3d/vendor/github.com/prometheus/prometheus/web/api/v1/api.go#L600). Unless we want to rewrite that handler we need to somehow "modify" `promql.QueryOptions`, IMO the `QueryEngine` interface is the only viable place. 

You're right, for MQE this can be implemented differently, see [this commit](https://github.com/grafana/mimir/pull/11848/commits/936bd5c8e646618cf3c70a4f65b649f0231b3021). The wrapper, is/was a way of unifying this logic for promql/mimir query engine paths.

aknuds1 commented on pkg/streamingpromql/query.go:
```diff
@@ -59,7 +61,7 @@ type Query struct {
 func (e *Engine) newQuery(ctx context.Context, queryable storage.Queryable, opts promql.QueryOpts, timeRange types.QueryTimeRange, originalExpression string) (*Query, error) {
 	if opts == nil {
-		opts = promql.NewPrometheusQueryOpts(false, 0)
+		opts = promql.NewPrometheusQueryOpts(false, 0, model.UTF8Validation)
```
@juliusmh Reading this again, `validationScheme` is defined on line 77. Could it be defined before line 63 and be used in the call to `NewPrometheusQueryOpts`? Just wondering whether it makes sense.

dimitarvdimitrov commented:
copying some feedback form slack on doing validation in the query engine:



> a user testing a recording or alerting rule as a query may not realise that their rule will later fail to write samples when evaluated by rulers
> […]
> or switches to legacy

so let’s say a user has had UTF8 for a bit, but for some reason (?) we want to turn it back off. Are we ok also breaking their query evaluations that work on the UTF8-ingested metrics?

Now their dashbaords will be broken and some alerts may not work until the blocker for enabling UTF8 is enabled. This sounds more intrusive and inconvenient than surprising them with failing rule evals. It makes something which was supposed to be async suddenly a top priority.

Always having UTF8 validation in the query engine looks safer and I agree with Bryan that it’s also simpler.

aknuds1 reviewed (dismissed):
There's a test change that I don't think should be made.

juliusmh commented on pkg/frontend/querymiddleware/request_validation_test.go:
```diff
@@ -208,9 +208,10 @@ func TestCardinalityQueryRequestValidationRoundTripper(t *testing.T) {
+			// non-legacy label name will be accepted
 			url:             cardinalityLabelValuesPathSuffix + "?label_names[]=\\xbd\\xb2\\x3d\\xbc\\x20\\xe2\\x8c\\x98",
-			expectedErrType: apierror.TypeBadData,
+			expectedErrType: "",
```
I think the test was wrong due to escape backslashes. I changed the case to test non-utf8 label name as the comment suggests. Also added a test case here ensuring regular utf8 names pass.

dimitarvdimitrov commented on go.mod:
i assume you'd want to merge the two PRs linked in the description before merging _this_ PR. Don't forget to do that :)

aknuds1 replied:
The Alertmanager PR is merged, only the mimir-prometheus PR remains. We will merge that first :)

dimitarvdimitrov commented on pkg/distributor/distributor.go:
```diff
@@ -1099,15 +1093,22 @@ func (d *Distributor) prePushRelabelMiddleware(next PushFunc) PushFunc {
 			ts := req.Timeseries[tsIdx]
 
-			if mrc := d.limits.MetricRelabelConfigs(userID); len(mrc) > 0 {
+			if mrcs := d.limits.MetricRelabelConfigs(userID); len(mrcs) > 0 {
```
nit to name `mrcs` something intelligible like `relabelConfigs`

aknuds1 replied:
@juliusmh Can you address the nitpick?

dimitarvdimitrov commented on pkg/distributor/distributor.go:
```diff
@@ -1099,15 +1093,22 @@ func (d *Distributor) prePushRelabelMiddleware(next PushFunc) PushFunc {
 
-			if mrc := d.limits.MetricRelabelConfigs(userID); len(mrc) > 0 {
+			if mrcs := d.limits.MetricRelabelConfigs(userID); len(mrcs) > 0 {
+				for _, mrc := range mrcs {
```
this currently runs for every incoming sample; can we make a change somewhere where the relabel configs are loaded so we set the validation scheme once instead of on every ingested sample for every relabel config?

juliusmh replied:
Was thinking about this when implementing. I would place this check in util/limits.validate, but the validation scheme doesn't have a yaml parameter (yet). 

I moved it up so it's not run for each series (but still for each sample). Do you have any other suggestion? @aknuds1

dimitarvdimitrov commented on pkg/distributor/validate_test.go:
```diff
@@ -236,10 +262,34 @@ func TestValidateLabels(t *testing.T) {
+		},
 	} {
-		err := validateLabels(s, cfg, userID, "custom label", mimirpb.FromMetricsToLabelAdapters(c.metric), c.skipLabelNameValidation, c.skipLabelCountValidation, cast, ts)
+		err := validateLabels(s, c.config, userID, "custom label", mimirpb.FromMetricsToLabelAdapters(c.metric), c.skipLabelNameValidation, c.skipLabelCountValidation, cast, ts)
```
i'm tempted to add more test cases for utf8 config. I think a lot of the existing cases would work in both validation schemes. Maybe have a slice of schemes to run each test cases in? or have two expected errors - one for legacy validation scheme and one for utf8

dimitarvdimitrov commented on pkg/distributor/validate_test.go:
```diff
@@ -249,7 +299,7 @@ func TestValidateLabels(t *testing.T) {
 			# HELP cortex_discarded_samples_total The total number of samples that were discarded.
 			# TYPE cortex_discarded_samples_total counter
-			cortex_discarded_samples_total{group="custom label",reason="label_invalid",user="testUser"} 1
+			cortex_discarded_samples_total{group="custom label",reason="label_invalid",user="testUser"} 2
```
nitpick to create this count out of the test cases instead of hardcoding it here. e.g. counting the number of expected errors. Would be amazing if you can do the same below; not sure how hard it would be

juliusmh replied:
I refactored the whole test to run each case for legacy and utf8 and compute the desired metrics using the test cases. This wasn't hard, but lengthy.

dimitarvdimitrov commented on pkg/frontend/querymiddleware/request_validation_test.go:
```diff
@@ -209,9 +209,13 @@ func TestCardinalityQueryRequestValidationRoundTripper(t *testing.T) {
 		{
 			// non-utf8 label name will be rejected even when we transition to UTF-8 label names
-			url:             cardinalityLabelValuesPathSuffix + "?label_names[]=\\xbd\\xb2\\x3d\\xbc\\x20\\xe2\\x8c\\x98",
+			url:             cardinalityLabelValuesPathSuffix + "?label_names[]=\xbd\xb2\x3d\xbc\x20\xe2\x8c\x98",
```
why was the previous url still rejected even though it was ascii-only? was it because of the slashes? let's keep it as a test case in addition to the real UTF8 case you added/fixed 🤷

dimitarvdimitrov replied:
i don't see any test cases for invalid strings

aknuds1 replied:
On second look, you are right @dimitarvdimitrov. Could you add a case @juliusmh?

juliusmh replied:
I don't really understand this comment. We have cases for
Invalid names: `\xbd\xb2\x3d` , valid legacy/utf8 name: `foo` and valid utf8 name: `some.label`

aknuds1 replied:
Ah yes, I wonder if I was looking at an outdated revision.

dimitarvdimitrov commented on pkg/frontend/querymiddleware/request_validation_test.go:
```diff
@@ -209,9 +209,13 @@ func TestCardinalityQueryRequestValidationRoundTripper(t *testing.T) {
 		{
 			// non-utf8 label name will be rejected even when we transition to UTF-8 label names
-			url:             cardinalityLabelValuesPathSuffix + "?label_names[]=\\xbd\\xb2\\x3d\\xbc\\x20\\xe2\\x8c\\x98",
+			url:             cardinalityLabelValuesPathSuffix + "?label_names[]=\xbd\xb2\x3d\xbc\x20\xe2\x8c\x98",
```
also can you make the smae changes to `TestMetricsQueryRequestValidationRoundTripper`  and `TestLabelsQueryRequestValidationRoundTripper`? maybe also in `TestPrometheusCodec_DecodeEncodeMultipleTimes_Labels`

dimitarvdimitrov replied:
oh, i'm surprised we don't do any validation on this request? do we delegate it to prometheus?

juliusmh replied:
As far as I can see, the request is not validated in `querymiddleware`. It looks like prometheus' `web/api/v1.API` already does validation on the query name, for label values, label names and series endpoints.

dimitarvdimitrov commented on pkg/mimir/promexts.go:
is there anything which depended on legacy validation in this package? The closes i could find is the runtime config, but i couldn't see how exactly

aknuds1 replied:
@dimitarvdimitrov I believe I just configured `model.LegacyValidation` in every package `init` function that I found necessary, when upgrading the Prometheus version where `model.NameValidationScheme` was introduced. To ensure that we didn't suddenly change the name validation scheme in Mimir.

dimitarvdimitrov replied:
why was it necessary here? we only turned it off, but didn't parametarize it. so if something depended on this init function it's now got the validation flipped

aknuds1 replied:
When working on this PR, we figured out what depends on `model.NameValidationScheme` by deleting that global, and working our way backwards.

dimitarvdimitrov replied:
but like you can't just make this change and hope it works 😄 i checked the package and i couldn't find anything which either directly or transitively needs validation

dimitarvdimitrov reviewed (commented):
i can't find any blockers. Maybe it would have been easier to get this through if you took it package by package

dimitarvdimitrov commented:
oh, also the mimirtool changes i think need a changelog entry because they're a change in behaviour

aknuds1 commented on pkg/mimirtool/commands/analyse_rulefiles.go:
```diff
@@ -22,8 +22,8 @@ type RuleFileAnalyzeCommand struct {
 func (cmd *RuleFileAnalyzeCommand) run(_ *kingpin.ParseContext) error {
-
-	output, err := AnalyzeRuleFiles(cmd.RuleFilesList)
+	// TODO: Get scheme from CLI flag.
```
@dimitarvdimitrov The idea is to leave the configurability (CLI flags etc) to the follow-up PR. The TODOs are for the next PR :)

aknuds1 commented on pkg/querier/cardinality_analysis_handler_test.go:
```diff
@@ -706,8 +706,8 @@ func TestLabelValuesCardinalityHandler_ParseError(t *testing.T) {
 				expectedErrorMessage: "'label_names[]' param is required",
 			},
 			"label_names param is invalid": {
-				url:                  "/label_values?label_names[]=olá",
```
Yes, it does become valid, because the query path validation scheme becomes UTF-8 instead of legacy. This makes me realize that the latter is actually a behavioral change (although something we want I think, from our discussions).

dimitarvdimitrov commented on pkg/mimirtool/commands/rules.go:
aha i see. it's confusing that the tests run with utf8 validation, but the main code runs with legacy validation

aknuds1 replied:
Yeah, we will be introducing support for UTF-8 validation in a follow-up.

juliusmh commented:
Given this was a bit chaotic, quick summary of what was addressed:
- refactored distributor's validation tests
- added more tests to `MetricsQueryRequestValidationRoundTripper`


and what is still open:
- [Here](https://github.com/grafana/mimir/pull/11848#discussion_r2225265876), I don't see which case is missing.
- [Here](https://github.com/grafana/mimir/pull/11848#discussion_r2225277677), should I add labelMatcher/selector validation to `LabelsQueryRequestValidationRoundTripper`? We didn't do that previously.
- [Here](https://github.com/grafana/mimir/pull/11848#discussion_r2225237252), Move this relabel config modification to a better place. Do you have any ideas where that could be?
- I added some Changelog messages, but needs refining

tacole02 commented on CHANGELOG.md:
```diff
@@ -4,6 +4,8 @@
 
 ### Grafana Mimir
 
+* [CHANGE] Query-frontend: Add support for utf8 label/metric names in `/api/v1/cardinality/{label_values|label_values|active_series}` endpoints. #11848.
+* [CHANGE] Querier: Add support for utf8 metric names, support utf8 label names in `label_join`, `label_replace` and `count_values` PromQL functions. #11848.
```
```suggestion
* [CHANGE] Query-frontend: Add support for UTF-8 metric and label names in `/api/v1/cardinality/{label_values|label_values|active_series}` endpoints. #11848.
* [CHANGE] Querier: Add support for UTF-8 metric and label names in `label_join`, `label_replace`, and `count_values` PromQL functions. #11848.
```
